src/bsoncxx/test/bson_serialization.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/document/value.hpp>
+#include <bsoncxx/string/to_string.hpp>
+#include <bsoncxx/test_util/catch.hh>
+#include <bsoncxx/types/bson_value/view.hpp>
+
+using namespace bsoncxx;
+
+using builder::basic::kvp;
+using builder::basic::make_array;
+using builder::basic::make_document;
+
+namespace {
+
+namespace test {
+
+struct Person {
+    std::string first_name;
+    std::string last_name;
+    int age;
+};
+
+// to_bson serializer function that can take a document::value&
+void to_bson(const Person& person, bsoncxx::document::value& bson_object) {
+    bson_object = make_document(kvp("first_name", person.first_name),
+                                kvp("last_name", person.last_name),
+                                kvp("age", person.age));
+}
+
+void from_bson(Person& person, const bsoncxx::document::view& bson_object) {
+    bson_object["first_name"].get_value().to_field(person.first_name);
+    bson_object["last_name"].get_value().to_field(person.last_name);
+    bson_object["age"].get_value().to_field(person.age);
+}
+
+class Car {
```

Comment:
I think the `Person` class alone provides enough evidence that this feature works as intended. I don't think much is gained from testing with additional classes/structs, aside from more code to maintain. Since this feature will just check that some object of type T has the function _declarations_ above reachable from the call sites in `value.hpp` we shouldn't need to worry about how the _definitions_ of those function will look.

Reply:
Yup, good point! I'll remove them!

The original intention of creating separate classes/structs was to make sure the serialization could be done with either a class or a struct, but this was probably more for my own personal testing. It was also intended that the other classes would later be used as nested objects for future testing (but this is irrelevant now).